toolkit/scripts/precache.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Usage: precache.sh </path/to/snapshot.json> <cache_working_dir> <cache_dir> <downloaded_files.txt> <repo_base_urls>...
+snapshot_path="$1"
+[[ -z "$snapshot_path" ]] && echo "snapshot_path is required" && exit 1
+shift
+
+cache_working_dir="$1"
+[[ -z "$cache_working_dir" ]] && echo "cache_working_dir is required" && exit 1
+shift
+
+cache_dir="$1"
+[[ -z "$cache_dir" ]] && echo "cache_dir is required" && exit 1
+shift
+
+downloaded_files="$1"
+[[ -z "$downloaded_files" ]] && echo "downloaded_files is required" && exit 1
+shift
+
+base_urls=("$@")
+[[ ${#base_urls[@]} -eq 0 ]] && echo "base_urls is required" && exit 1
+
+mkdir -p "$cache_working_dir"
+mkdir -p "$cache_dir"
+rm -f "$downloaded_files"
+
+. /etc/os-release
+if [[ "$ID" == "mariner" ]]; then
+    REPOQUERY_OS_ARGS="-y -q --disablerepo=*"
+else
+    REPOQUERY_OS_ARGS="--show-duplicates --tempcache"
+fi
+
+skipped_file="$cache_working_dir/skipped.txt"
+repo_summary_file="$cache_working_dir/repo.txt"
+
+# Init the cache
+mkdir -p "$cache_working_dir"
+mkdir -p "$cache_dir"
+rm -f "$downloaded_files"
+rm -f "$skipped_file"
+
+# For each base url, pull the meta data and write it to the repo summary file
+echo "" > $repo_summary_file
+for base_url in "${base_urls[@]}"; do
+    repo_unique="repo-$(echo $(((RANDOM%99999)+10000)))"
+    repo_name="mariner-precache-$repo_unique"
+    echo "Querying repo $base_url via 'repoquery $REPOQUERY_OS_ARGS --repofrompath=$repo_name,$base_url -a --qf="%{location}" >> $repo_summary_file'"
+    if [[ "$ID" == "mariner" ]]; then
```

Comment:
Since we now have some time before we need the performance upgrades, could we move this logic to be more aligned with the rest of our tooling to a Go tool? Either as an extension to our current fetcher or something along the lines of `specarchchecker` or `rpmssnapshot`?

Reply:
Sure, I can do that if we don't mind waiting on this. The script is getting really ugly anyways.

Reply:
Ok, and now in golang.

Reply:
About waiting for this: that was my worry too, so I discussed our schedule with @jslobodzian and we're fine to have this finished in September. The main point was you were able to prove we can run faster and now we're given time to implement it in a cleaner way.